lib/compilation_engine.rb

```diff
@@ -57,52 +68,192 @@ def compile_subroutine
     end
 
-    output_token
+    output_token # ;
+
     @output.puts("</varDec>")
   end
 
+  def compile_statements
+    @output.puts("<statements>")
+
+    until !@tokenizer.has_more_tokens? || is_symbol_token_and_equal?("}")
+      case @tokenizer.key_word
+      when :RETURN
+        compile_return
+      when :LET
+        compile_let
+      when :WHILE
+        compile_while
+      when :IF
+        compile_if
+      when :DO
+        compile_do
+      end
+    end
+
+    @output.puts("</statements>")
+  end
+
+  def compile_return
+    @output.puts("<returnStatement>")
+
+    output_token # return
+
+    unless is_symbol_token_and_equal?(";")
+      compile_expression
+    end
+
+    output_token # ;
+
+    @output.puts("</returnStatement>")
+  end
+
+  def compile_let
+    @output.puts("<letStatement>")
+    output_token # let
+    output_token # varName
+
+    if is_symbol_token_and_equal?("[")
+      output_token # [
+      compile_expression
+      output_token # ]
+    end
+
+    output_token # =
+
+    compile_expression # expression
+
+    output_token # ;
+    @output.puts("</letStatement>")
+  end
+
+  def compile_while
+    @output.puts("<whileStatement>")
+    output_token # while
+
+    output_token # (
+    compile_expression
+    output_token # )
+
+    output_token # {
+    compile_statements
+    output_token # }
+    @output.puts("</whileStatement>")
+  end
+
+  def compile_if
+    @output.puts("<ifStatement>")
+    output_token # if
+
+    output_token # (
+    compile_expression
+    output_token # )
+
+    output_token # {
+    compile_statements
+    output_token # }
+
+    if is_keyword_token_and_equal?(:ELSE)
+      output_token # else
+      output_token # {
+      compile_statements
+      output_token # }
+    end
+
+    @output.puts("</ifStatement>")
+  end
+
+  def compile_expression
+    @output.puts("<expression>")
+    compile_term
+    @output.puts("</expression>")
+  end
+
+  def compile_expression_list
+    @output.puts("<expressionList>")
+
+    unless is_symbol_token_and_equal?(")")
+      compile_expression
+
+      while is_symbol_token_and_equal?(",")
+        output_token # ,
+        compile_expression
+      end
+    end
+
+    @output.puts("</expressionList>")
+  end
+
+  def compile_do
+    @output.puts("<doStatement>")
+    output_token # do
+    compile_subroutine_call
+    @output.puts("</doStatement>")
+  end
+
+  def compile_term
+    @output.puts("<term>")
+    output_token # int / str / keyword / identifier / subroutine call / expression / unary op
+    @output.puts("</term>")
+  end
+
   private
 
   def compile_subroutine_body
     @output.puts("<subroutineBody>")
 
-    output_token
+    output_token # {
 
-    advance
-    if tokenizer.token_type == :KEYWORD && tokenizer.key_word == :VAR
+    while is_keyword_token_and_equal?(:VAR)
       compile_var_dec
     end
 
-    advance_and_output_token
+    if tokenizer.token_type == :KEYWORD
+      compile_statements
+    end
```

Comment:
In the same way as we removed the conditional around `compile_statements` in `#compile_while` and `#compile_if` (3d538c471b6355cc3d216846979c581137162af2), I think we should do the same thing here:

```suggestion
    compile_statements
```

Again we’d need to update a few unit tests to include the empty `<statements></statements>` tag in the expected output, but I think that’s more consistent with what we’ve done elsewhere.